pkg/metrics/metrics.go

```diff
@@ -259,16 +275,27 @@ func (c *client) WriteMetric(ctx context.Context, name string, count int64) erro
 //	    // handle error
 //	  }
 //	}()
-func (c *client) WriteMetricAsync(ctx context.Context, name string, count int64) func() error {
-	errCh := make(chan error, 1)
+func (c *client) WriteMetricAsync(ctx context.Context, name string, count int64) {
+	if c.OptOut {
+		return
+	}
+	c.AsyncRunners.Add(1)
 	go func() {
-		defer close(errCh)
-		errCh <- c.WriteMetric(ctx, name, count)
+		if err := c.WriteMetric(ctx, name, count); err != nil && c.ErrorHandler != nil {
+			c.ErrorHandler(ctx, err)
+		}
+		c.AsyncRunners.Done()
 	}()
+}
 
-	return func() error {
-		return <-errCh
+// Close blocks for all async Metrics to finish. Operations after Close()
+// returns will be noops.
+func (c *client) Close() {
+	if c.OptOut {
+		return
 	}
+	c.AsyncRunners.Wait()
+	c.OptOut = true
```

Comment:
Is there a race condition here? AFAICT the writer and readers of the OptOut variable aren't synchronized in any way, and there always needs to be synchronization when mutable memory is shared.

Reply:
Yes there is. I was thinking that it wouldn't matter to much because the worst case is a leaked goroutine, but I guess it makes sense to ensure that can't happen.

Reply:
I have added a solution that I think is correct (despite tryRLock being discourage by go docs).